### PR TITLE
220 df index sort mgtfs

### LINKED ISSUE #220: multigtfs df attribute row sort
Also applies to summarise_trips, row and column sort should be consistent with that of GtfsInstance:
row sort on day is: ['friday', 'friday', 'monday', 'monday', 'saturday', 'saturday', 'sunday', 'sunday', 'thursday', 'thursday', 'tuesday', 'tuesday', 'wednesday', 'wednesday']
column sort is: ['day', 'route_type', 'route_count_max', 'route_count_mean', 'route_count_median', 'route_count_min']
GtfsInstance column sort is ['day', 'route_count_max', 'route_count_mean', 'route_count_median', 'route_count_min', 'route_type']

_Originally posted by @r-leyshon in https://github.com/datasciencecampus/transport-network-performance/pull/207#discussion_r1410359868_

### PR DESCRIPTION
**Pull request by cbrowning04**
## Description

Adds additional sorting options for mgtfs summaries

Fixes #220 

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How Has This Been Tested?

Test configuration details:
* OS: Windows 10
* Python version: 3.9.13
* Java version: N/A
* Python management system: MiniConda

## Checklist:

- [x] My code follows the intended structure of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: src/transport_performance/gtfs/multi_validation.py [hunk 1]
```diff
@@ -405,6 +414,12 @@ def _summarise_core(
             trips/routes for each date. When False, summ_ops becomes useless,
             and should therefore nothing should be passed when calling this
             function (so it remains as the default). Defaults to False.
+        sort_by : str, optional
```
Comment (SergioRec) [inline src/transport_performance/gtfs/multi_validation.py]: Not sure if I understand this argument.

It only accepts `["days", "route_type"]`. However, the function is always sorting by day, so the argument "days" doesn't actually do anything. If "route_type" is included, then it sorts by this in addition to "days".

What is the reason for having a "days" option if this doesn't change the behaviour of the function?

Also why adding the extra processing step of sorting by day when this is then overridden by sorting by route type? You mention in a comment that `route_type` sorts are more organised, but I'm not sure this is the case.

Perhaps the following would be better:

```

if sort_by == "days":
    trip_counts = self.instances[0]._order_dataframe_by_day(trip_counts)
elif sort_by == "route_type":
    trip_counts = self.instances[0].sort_values("route_type")

```

File: src/transport_performance/gtfs/multi_validation.py [hunk 2]
```diff
@@ -476,14 +498,23 @@ def _summarise_core(
             column.replace("amin", "min").replace("amax", "max")
             for column in trip_counts.columns.values
         ]
+        # sorting
+        # NOTE: More convoluted sorts don't work well with the custom
+        # _order_dataframe_by_day func.
+        # Always sort by day so that route_type sorts are more organised
         trip_counts = self.instances[0]._order_dataframe_by_day(trip_counts)
+        if sort_by == "route_type":
```
Comment (SergioRec) [inline src/transport_performance/gtfs/multi_validation.py]:
```
gtfs.summarise_routes(sort_by='days').equals(gtfs.summarise_routes(sort_by='route_type'))
```

This is true even when loading a gtfs with more than two route types.

Is this only useful when using `to_days = True`? Not sure if any changed is needed, but maybe some documentation could help.

File: src/transport_performance/gtfs/multi_validation.py [hunk 3]
```diff
@@ -476,14 +498,23 @@ def _summarise_core(
...
+        if to_days:
+            trip_counts = self._summary_col_sorter(trip_counts)
         return trip_counts
 
     def summarise_trips(
         self,
         summ_ops: list = [np.min, np.max, np.mean, np.median],
         return_summary: bool = True,
         to_days: bool = False,
+        sort_by: Union[str, list] = "days",
```
Comment (SergioRec) [inline src/transport_performance/gtfs/multi_validation.py]: This type hint (and docstring below) should be `str` only, as function `_summarise_core` using this argument only accepts `str`.

File: src/transport_performance/gtfs/multi_validation.py [hunk 4]
```diff
@@ -476,14 +498,23 @@ def _summarise_core(
...
             for column in trip_counts.columns.values
         ]
+        # sorting
+        # NOTE: More convoluted sorts don't work well with the custom
+        # _order_dataframe_by_day func.
+        # Always sort by day so that route_type sorts are more organised
         trip_counts = self.instances[0]._order_dataframe_by_day(trip_counts)
+        if sort_by == "route_type":
+            trip_counts = trip_counts.sort_values("route_type")
+        if to_days:
```
Comment (SergioRec) [inline src/transport_performance/gtfs/multi_validation.py:496]: Note that if `to_days` is set to `True`, the index after sorting will be messed up. This is nitpicking, but perhaps we could add a `reset_index` somewhere.

File: src/transport_performance/gtfs/multi_validation.py [hunk 5]
```diff
@@ -379,12 +379,21 @@ def filter_to_bbox(
             self._raise_empty_feed_error(bbox)
         return None
 
+    def _summary_col_sorter(self, df: pd.DataFrame) -> pd.DataFrame:
```
Comment (SergioRec) [inline src/transport_performance/gtfs/multi_validation.py:382]: No unit tests for this, would it be worth adding them?

**Review by SergioRec — CHANGES_REQUESTED** [review]
Thanks for your work on this @CBROWN-ONS. All seems to work as intended. There's only a couple of actions to take:
- Unit tests for `_summary_col_sorter`.
- Corrections to type hint and docstring for `summarise_trips`.

The rest are suggestions/clarifications or maybe misunderstandings on my part. Happy to discuss any of the points!

File: src/transport_performance/gtfs/multi_validation.py (hunk 1 above)
Comment (cbrowning04) [reply src/transport_performance/gtfs/multi_validation.py]: I've addressed this issue by restructuring the parameter. Good spot :)

File: src/transport_performance/gtfs/multi_validation.py (hunk 3 above)
Comment (cbrowning04) [reply src/transport_performance/gtfs/multi_validation.py]: Fixed by restructuring of param

File: src/transport_performance/gtfs/multi_validation.py (hunk 4 above)
Comment (cbrowning04) [reply src/transport_performance/gtfs/multi_validation.py:496]: Good shout. Added this

File: src/transport_performance/gtfs/multi_validation.py (hunk 2 above)
Comment (cbrowning04) [reply src/transport_performance/gtfs/multi_validation.py]: Done 👍

File: src/transport_performance/gtfs/multi_validation.py (hunk 5 above)
Comment (cbrowning04) [reply src/transport_performance/gtfs/multi_validation.py:382]: Going to say no on this as it is an internal function. It is also tested through `test__summarise_core()`  through dataframe equality checks, including column ordering.

**Comment by cbrowning04** [comment]
Hi @SergioRec . I've made some changes based on your comments 😄 .

File: src/transport_performance/gtfs/multi_validation.py (hunk 4 above)
Comment (SergioRec) [reply src/transport_performance/gtfs/multi_validation.py:496]: I've changed the `reset_index` to the return as this wasn't fixing the issue when `to_days` was `True`. Let me know if you're not happy with this!

**Review by SergioRec — APPROVED** [review]
Thanks for addressing all my comments @CBROWN-ONS.

I've made a small change to the location of the `reset_index` in `_summarise_core`. Could you please check if you're happy with it. Other than that, everything else works and this PR can be merged. Thanks!

**Comment by cbrowning04** [comment]
Looks good to me @SergioRec . Happy for this to be merged

**Pull request merged**